File: LeetCode_Java/49_groupAnagrams.java
import java.util.*;

class GroupAnagrams{
    public List<List<String>> groupAnagrams(String[] strs){
        Map<String,List<String>> anagramMap = new HashMap<>();
        
        for(String str : strs){
            int[] countArray = new int[26];

            for(char c : str.toCharArray()){
                countArray[c - 'a']++;
            }

            String key = Arrays.toString(countArray);

            if(!anagramMap.containsKey(key)){
                anagramMap.put(key, new ArrayList<>());
            }

            anagramMap.get(key).add(str);
        }
        return new ArrayList<>(anagramMap.values());
    }

    public static void main(String[] args){

        GroupAnagrams solution = new GroupAnagrams();
        String[] strs = {"eat", "tea", "tan", "ate", "nat", "bat"};

        List<List<String>> result = solution.groupAnagrams(strs);

        for(List<String> group : result){
            System.out.println(group);
        }

    }
}

File: LeetCode_Python/49_groupAnagrams.py
from collections import defaultdict
from typing import List


class GroupAnagrams:
    def groupAnagrams(self, strs : List[str]) -> List[List[str]]:
        
        anagramMap = defaultdict(list)
        
        for str in strs:
            countArray = [0] * 26
            
            for c in str:
                countArray[ord(c) - ord('a') ] += 1
                
            anagramMap[tuple(countArray)].append(str)
            
        return anagramMap.values()
    
if __name__ == "__main__":
    solution = GroupAnagrams()
    
    strs = ["eat","tea","tan","ate","nat","bat"]
    
    print(solution.groupAnagrams(strs))